### ISSUE TITLE
Known install issues

### ISSUE THREAD

**Issue by Anniepoo** [Anniepoo/swiplsitedesign#50]
This is a place to collect known problems with the install process.
There's going to be a troubleshooting section in the appropriate place for each type.

Note to myself - need to place these on install or testing install pages as appropriate

Issues I know of:
     Building on linux - 
- people fail to build packages, and the system comes up halfway.
- People have trouble with the graphics? Not sure whats up with that.
- accounts and install
  
  Windows install - 
- Path length restriction  
- Is there an issue with install on non admin accounts?

Ideally we'd improve the actual install process to deal with these issues, but until then we need to provide troubleshooting

Jan - You probably know better what sort of trouble people get into, over to you to add ones you know about.

**Comment by JanWielemaker** [comment]
It is a difficult and very dynamic issue.   Problems come and go in a matter of months.
Given different OSes and the many dependencies of the system, this is probably hard to avoid.

Having wiki pages is more or less the only sensible way out.  The general direction though
is to solve the problems once we get an idea what the real problem is and if we can find a
way around.

Might be nice if we can find a way to run some predicate that checks for common problems
with the installation (provided the basic stuff works) and points to a web page about them.

At the end of the wizard, we should be able to bring up a page/link with the relevant info,
possibly two: a short one that is included (dynamically) in the download page with a link
to a more elaborate trouble-shooting list.

**Comment by Anniepoo** [comment]
no, at the end of the wizard you've installed. Info at that point isn't actionable.

**Comment by Anniepoo** [comment]
Jan - maybe we should talk on skype tonight and hammer this out.

**Comment by JanWielemaker** [comment]
On 01/27/2014 12:22 AM, Anne Ogborn wrote:

> Jan - maybe we should talk on skype tonight and hammer this out.

Ok.  I'm online for the next couple of hours.  I just checked out your
download branch.  Got it working to the point where I see three buttons.
Clicking them gives an error.  Not sure whether this is expected or I
missed something in the setup.  Reading through the changes give me some
clues though.

I'm not convinced by this direction.  I know about a lot of issues, but
_finding_ the file to download is hardly one of them and it is only three
clicks (download, stable/devel, file).

I see some things going wrong:
- People downloading the stable or 32-bit version while they need the
  devel (because they want to do fancy stuff) or 64 bit (because they
  have a lot of data).
- People downloading the MacOS version and having trouble with (notably)
  X11 or adjusting their PATH to run the command line tools.
- People downloading the source and failing to build it, notably
  failing to install the dependencies first and/or not using the
  build.template route and thus ending up with only the core.
- People trying to embed SWI-Prolog and running into all sorts of
  setup issues.  I guess it is hard to deal with that as the scenarios
  are endless.

Note that branching on OS first is not necessarily the right start.
Some people have the desire to run some program and have options wrt.
the OS.  Probably neither is stable/devel as it is now.  That is why
I propose to show the multi-dimensional choice all at once in a number
of radio buttons.  That way you can make the next choice yourself and
you see things being greyed out.  Then, there are three links:
[download], [download docs] and [license] and a (wiki) text explaining
what to do next.

I see most value in ability to show dedicated texts that can be a bit longer
and explain what one can expect and is supposed to do next that is specific
to the platform, devel/stable, source/binary.

So, if you are around, lets chat on skype.

```
Cheers --- Jan
```

**Comment by JanWielemaker** [comment]
Hi Anne,

Added a quick multi-dimensional set of notes as download.md to
swiplsitedesign.

```
Cheers --- Jan
```

**Comment by JanWielemaker** [comment]
Added some more notes to download.md.  Curious to hear what is wrong
with it :-)
I hear you thinking `too complicated'.  Given sensible preselection and
a layout
that makes where the download buttons stand out, I hope this is ok.

Note how I propose to deal with upgrading.  Note that we have all the
infrastructure
to generate changelogs automatically.  Could use a better presentation
though :-(

```
Cheers --- Jan
```

Going offline now.  Got to get up early tomorrow.

**Comment by Anniepoo** [comment]
Jan, if you're up, I'm on now, lets skpe.

My availability's otherwise going to be limited next few days
between work and moving.

**Comment by Anniepoo** [comment]
the checkin just had the windows branch working, and that only as far as first couple pages - definitely not finished, I just ran out of time.

Not much time now, I'm awake for a few minutes in middle of night, but -
I had the same thought about radio buttons/rules. I got as far as implementing some radio buttons that showed this sort of thing, and it was immediately obvious that it wasn't going to work visually. So I went to the wizard.

This isn't actually a change in the set of tasks, it's just separating them visually so
- it's not obvious everything else is changing when one thing changes (source of visual confusion).
- it's not obvious 

Yes, selecting proper download's not the hard part. Hard part is getting people to do the whole install process, etc.
- People downloading the source and failing to build it, notably
  failing to install the dependencies first and/or not using the
  build.template route and thus ending up with only the core.

Part of idea of wizard was to make sure all this happens.

I think looking at the instructions that come in various places would help. 

Anyway, yes, lets hammer out on skype

**Comment by JanWielemaker** [comment]
On 01/28/2014 01:34 PM, Anne Ogborn wrote:

> the checkin just had the windows branch working, and that only as far as
> first couple pages - definitely not finished, I just ran out of time.
> 
> Not much time now, I'm awake for a few minutes in middle of night, but -
> I had the same thought about radio buttons/rules. I got as far as
> implementing some radio buttons that showed this sort of thing, and it
> was immediately obvious that it wasn't going to work visually. So I went
> to the wizard.

I'm noty yet convinced.  Do you have the attempt around?

> Yes, selecting proper download's not the hard part. Hard part is getting
> people to do the whole install process, etc.
> - People downloading the source and failing to build it, notably
>   failing to install the dependencies first and/or not using the
>   build.template route and thus ending up with only the core.
> 
> Part of idea of wizard was to make sure all this happens.
> 
> I think looking at the instructions that come in various places would help.

Possibly.  Not easily though because the variations are endless and change
quite rapidly with changing OS versions (renaming packages, incompatible
packages) and Prolog versions (different dependencies, new work-arounds,
etc.)  The best we can do is probably adding a predicate or menu entry
`check_installation' that will verify that everything is present or point
what is missing with a link on more info

```
Cheers --- Jan
```

**Comment by Anniepoo** [comment]
no, don't have that code  - I just put in html to make the buttons, and when I saw that sea of buttons and imagined changing one of them and the others reconfiguring it was obvious.

 I think looking at the instructions that come in various places would help.
- I was thinking of the instructions for building on linux, which vary from place to place. I don't have time right now to dig them all up, but have been confused myself by this when building in the past.
  Additionally, would it make sense for the main build script to ask if you want to build packages and if so, go ahead and build them? That'd eliminate the didn't-build-packages problem.

**Comment by JanWielemaker** [comment]
![download](https://f.cloud.github.com/assets/3071146/2048684/99dcf304-8a4f-11e3-9274-84e8011127db.png)
Hi Anne,

On 01/29/2014 02:01 PM, Anne Ogborn wrote:

> no, don't have that code - I just put in html to make the buttons, and
> when I saw that sea of buttons and imagined changing one of them and the
> others reconfiguring it was obvious.

I don't give up that easily as I think it
1. Turns default download into a single click for many users (based on
     browser id).
2. Immediately shows you what the alternatives are

So, I decided to make a little mockup that generates an actual page from
a Prolog handler, but otherwise does nothing.  I've attached the 
screenshot, showing what it might look like on first open from a Windows
user.

A lot can be said about the details of the labels and the text below, 
but I do think that it looks acceptable and provides both brevity and
the (I think important) overview.  As long as we keep the item list
below your famous 8 ...

> I think looking at the instructions that come in various places would help.
> - I was thinking of the instructions for building on linux, which vary
>   from place to place. I don't have time right now to dig them all up,
>   but have been confused myself by this when building in the past.
>   Additionally, would it make sense for the main build script to ask
>   if you want to build packages and if so, go ahead and build them?
>   That'd eliminate the didn't-build-packages problem.

Yes.  That is a mess.  The main build script however does install all
packages that also come with the default binary for all platforms. 
Only, it skips things if it cannot find the dependencies.  I think
that all the outdated descriptions should vanish and that the build
script should print a summary at the end that explains the status
of the build.  For example (we can add wiki pages for each thing that
commonly fails):

# 

Build completed.  The following components failed to build:

```
- JPL  -- see http://www.swi-prolog.org/build/jpl.html
     - xpce -- see http://www.swi-prolog.org/build/xpce.html

Cheers --- Jan
```

**Comment by Anniepoo** [comment]
Your mockup didn't come through the github email system.

**Comment by JanWielemaker** [comment]
Edited.  There is a link at the top.
